File: Javascript/bubble_sort.js
/**
 * This file is part of [Algorithms]
 *
 * (c) 2024 [Nainik Mehta] <[nainikmehta25@gmail.com]>
 *
 * --------------------------------------------------
 *
 * @module app.v1.bubbleSort
 * @description Bubble sort is a simple sorting algorithm that repeatedly steps through the list, compares adjacent elements and swaps them if they are in the wrong order, and efficiently identifies the sorted elements with each successive pass through the list.
 * @author [Nainik Mehta] <[https://www.linkedin.com/in/nainik-mehta-25nk12/]>
 * @github - Nainik2509  <[https://github.com/Nainik2509/]>
 * @version 1.0.0
 *
 * --------------------------------------------------
 */

function bubbleSort(arr) {
  let len = arr.length; // Get the number of elements in the array
  let swapped;

  do {
    swapped = false; // This flag will monitor whether elements were swapped during the pass

    for (let i = 0; i < len - 1; i++) {
      // Loop through the array up to the second to last element
      if (arr[i] > arr[i + 1]) {
        // Compare current element to the next one
        // If the current element is greater, swap them
        [[arr[i], arr[i + 1]]] = [[arr[i + 1], arr[i]]];
        swapped = true; // Set the swapped flag to true indicating a swap occurred
      }
    }
    // Continue the outer loop until no swaps are made on a new pass, indicating the array is sorted
  } while (swapped);

  return arr; // Return the sorted array
}

// Example usage
console.log(bubbleSort([1, 4, 2, 3, 5, 75, 3])); // Output will be the sorted array

File: C++/Sorting/bubble_sort.cpp
#include <iostream>
using namespace std;

// Function to swap two elements using pointers
void swap(int *x, int *y)
{
    int temp = *x; // Temporary storage of the value at address x
    *x = *y;       // Assign the value at address y to address x
    *y = temp;     // Assign the value in temp (original x) to address y
}

// Implementation of the Bubble Sort algorithm
void BubbleSort(int array[], int size)
{
    int i, j;
    bool isSwapped;

    // Loop through all elements in the array
    for (i = 0; i < size - 1; i++)
    {
        isSwapped = false; // Flag to check if any swap happened

        // Inner loop to compare array elements
        for (j = 0; j < size - 1 - i; j++) // size-1-i, as the last i elements are already sorted
        {
            // If current element is greater than the next, swap them
            if (array[j] > array[j + 1])
            {
                swap(&array[j], &array[j + 1]);
                isSwapped = true; // Set the flag to true showing a swap occurred
            }
        }

        // If no two elements were swapped by inner loop, then the array is sorted
        if (isSwapped == false)
        {
            break; // Break out of the loop early
        }
    }
}

// Main function to run the Bubble Sort
int main()
{
    int arr[] = {2, 4, 3, 7, 6, 9}; // Array to be sorted
    int size = 6;                   // Number of elements in the array

    // Call the BubbleSort function
    BubbleSort(arr, size);

    // Printing the sorted array
    for (int i = 0; i < size; i++)
    {
        printf("%d ", arr[i]);
    }

    return 0;
}
